### ISSUE TITLE
SSL bypass not working

### ISSUE THREAD

**Issue by MartinOehler** [e2guardian/e2guardian#135]
Using e2guardian 3.4.0.3, I am unable to bypass SSL page blocks. 

The setup is using the mitm configuration as described  - everything is working for SSL and Non-SSL pages until the blanket block is activated in bannedsitelist:

```
#Blanket Block.  To block all sites except those in the
#exceptionsitelist and greysitelist files, remove
#the # from the next line to leave only a '**':
**
```

Unblocking of http (Non-SSL) pages still works, but when accessing a https URL like https://example.com the redirect URL

```
https://example.com/?GBYPASS=6C997BD76297CE0900CA61873B4CA70E1470226003
```

is shown correctly but after selecting the URL there is **another** DENY page showing up for unblocking the GBYPASS-URL. The target URL is then

```
https://example.com/?GBYPASS=6C997BD76297CE0900CA61873B4CA70E1470226003&GBYPASS=54FF8D5656ADF8A64F2191BC140CDC0D1470226007
```

and so on.

The log entry is:

```
[...]https://example.com/?GBYPASS=6C997BD76297CE0900CA61873B4CA70E1470226003 *DENIED* Blocked site: Walled Garden is on and the site is not available to you.
```

Is using the bypass mechanism for SSL not supported at all or might this be a bug?

Thanks,
Martin

**Comment by MartinOehler** [comment]
Here is the debug output of a non-working SSL bypass with mitm configuration

  http://pastebin.com/pwvah7j3

Perhaps this helps.

Martin

**Comment by MartinOehler** [comment]
The problem concerning the denial of the connect when using the GBYPASS URL is rooted in ````ConnectionHandler::hashedURL````, ````ConnectionHandler::hashedCookie```` and ````HTTPHeader::isBypassURL````. There is a mismatch in mitm mode, some hashes are calculated with http://example.com and some with https://example.com which resulted in different hashes.

I solved this by calculating and checking the hashes without using the protocol. So far so good, the hashes are matching but now e2guardian is returning the http page for an https request after using the GBYPASS URL. :-(

The bypassing itself works now, when entering the https URL afterwards the https page is correctly displayed.

Martin

**Comment by MartinOehler** [comment]
After returning the referer as redirect location for the above case, the https page is displayed. The redirect returned the http page that is used inside the tunnel as redirect location in e2guardian.

I'll test it for two or three days and provide a cleaned up patch.

Martin

**Comment by philipianpearce** [comment]
Hi, 

I have had a chance to have a look at this today. 

If you change this line in ConnectHandler.cpp:- 

if (header.isScanBypassURL(&url, o.fg[filtergroup]->magic.c_str(), clientip.c_str())) { 
to 
if (header.isScanBypassURL(&logurl, o.fg[filtergroup]->magic.c_str(), clientip.c_str())) { 

it should fix the problem which was that the deny function is called with logurl and isScanBypassURL with url. 

logurl is the url as originally would have been called in browser (in the case of ssl with https: prefix) 

url is the url being worked on (in the case of ssl mitm will have http: prefix as it is being read inside the ssl tunnel) 

With http: url and logurl will generally be the same. With https mitm they will always be different. 

I did not pick up this problem when doing the original MITM coding/testing as we do not use the bypass function ourselves and so it was not checked as part of our testing process. 

Regards 

Philip Pearce 

----- Original Message -----

**Comment by MartinOehler** [comment]
This is of course the cleaner approach. But after using only your one-line change I got the original behaviour (GBYPASS URL as blocked site). Using your proposal of using logurl instead of url I was successful after patching ````ConnectionHandler.cpp```` in some more places:

1.  After modification of the ````isBypassURL```` lines (url->logurl) I got the http page instead of the https page.
2.  After additional modification of 
````
loc += header.getUrl(true); -> loc += header.getLogUrl(true);
````
  it works as expected. Attached my [patch](https://github.com/e2guardian/e2guardian/files/406583/Connectionhandler.patch.txt) for review.

Thanks,
Martin

**Comment by eterry1388** [comment]
Seeing this issue as well.

**Comment by fredbcode** [comment]
Philip, this patch seems good to me what do you think ?
If you agree I will push this to develop

**Comment by philipianpearce** [comment]
Fred, Yes patch looks good, agree to push. 

Regards 

Philip 

----- Original Message -----

**Comment by fredbcode** [comment]
Ok fixed

**Comment by fcruzcoelho** [comment]
Good afternoon,

I am with this problem this correction is already in the last version of e2guardian?

My version is:
E2guardian 3.4.0.3

If not, how would I apply it?

Thank you

**Comment by fredbcode** [comment]
Sorry this fix is only for the next release 3.5, I hope we can release in a couple of weeks.
In our forum there is a test package for debian

Le 26 janvier 2017 17:39:41 GMT+01:00, fcruzcoelho <notifications@github.com> a écrit :
>Good afternoon,
>
>I am with this problem this correction is already in the last version
>of e2guardian?
>
>My version is:
>E2guardian 3.4.0.3
>
>If not, how would I apply it?
>
>Thank you
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/e2guardian/e2guardian/issues/135#issuecomment-275438411

-- 
Envoyé de mon appareil Android avec K-9 Mail. Veuillez excuser ma brièveté.

**Comment by fredbcode** [comment]
Fixed in 3.5.0
